src/bun.js/api/bun/h2_frame_parser.zig

```diff
@@ -0,0 +1,2440 @@
+                this.state = .CLOSED;
+                this.client.endStream(this, .CANCEL);
+                this.client.dispatchWithExtra(.onAborted, JSC.JSValue.jsNumber(this.id), reason);
+            }
+        }
+
+        pub fn deinit(this: *Stream) void {
+            if (this.signal) |signal| {
+                this.signal = null;
+                signal.detach(this);
+            }
+        }
+    };
+
+    const HeaderValue = struct {
+        name: []const u8,
+        value: []const u8,
+        next: usize,
+    };
+
+    pub fn decode(this: *H2FrameParser, header_buffer: *[MAX_HPACK_HEADER_SIZE]u8, src_buffer: []const u8) !HeaderValue {
+        var xhdr: lshpack.lsxpack_header = .{};
+
+        lshpack.lsxpack_header_prepare_decode(&xhdr, header_buffer.ptr, 0, header_buffer.len);
+        var start = @intFromPtr(src_buffer.ptr);
+        var src = src_buffer.ptr;
+        if (lshpack.lshpack_dec_decode(&this.decoder, &src, @ptrFromInt(start + src_buffer.len), &xhdr) != 0) {
+            return error.UnableToDecode;
+        }
+        const name = lshpack.lsxpack_header_get_name(&xhdr);
+        if (name.len == 0) {
+            return error.EmptyHeaderName;
+        }
+        return .{
+            .name = name,
+            .value = lshpack.lsxpack_header_get_value(&xhdr),
+            .next = @intFromPtr(src) - start,
+        };
+    }
+
+    pub fn encode(this: *H2FrameParser, header_buffer: *[MAX_HPACK_HEADER_SIZE]u8, dst_buffer: []const u8, name: []const u8, value: []const u8, never_index: bool) !usize {
+        var xhdr: lshpack.lsxpack_header = .{ .indexed_type = if (never_index) 2 else 0 };
+        const size = name.len + value.len;
+        if (size > MAX_HPACK_HEADER_SIZE) {
+            return error.HeaderTooLarge;
+        }
+
+        @memcpy(header_buffer[0..name.len], name);
+        @memcpy(header_buffer[name.len..size], value);
+        lshpack.lsxpack_header_set_offset2(&xhdr, header_buffer.ptr, 0, name.len, name.len, value.len);
+        if (never_index) {
+            xhdr.indexed_type = 2;
+        }
+
+        var start = @intFromPtr(dst_buffer.ptr);
+        log("encode xhdr name {} {} val {} {}", .{ xhdr.name_offset, xhdr.name_len, xhdr.val_offset, xhdr.val_len });
+        const ptr = lshpack.lshpack_enc_encode(&this.encoder, dst_buffer.ptr, @ptrFromInt(start + dst_buffer.len), &xhdr);
+        const end = @intFromPtr(ptr) - start;
+        if (end > 0) {
+            return end;
+        }
+        return error.UnableToEncode;
+    }
+
+    /// Calculate the new window size for the connection and the stream
+    /// https://datatracker.ietf.org/doc/html/rfc7540#section-6.9.1
+    fn ajustWindowSize(this: *H2FrameParser, stream: ?*Stream, payloadSize: u32) void {
+        this.usedWindowSize += payloadSize;
+        if (this.usedWindowSize >= this.windowSize) {
+            var increment_size: u32 = WINDOW_INCREMENT_SIZE;
+            var new_size = this.windowSize + increment_size;
```

Comment:
let's make sure this can't overflow:
```suggestion
            var new_size = this.windowSize +| increment_size;
```